src/router.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright (C) 2019 Alibaba Cloud. All rights reserved.
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+
+use crate::{MediaType, Method, Request, Response, StatusCode, Version};
+
+pub use crate::common::RouteError;
+
+/// An HTTP endpoint handler interface
+pub trait EndpointHandler<T>: Sync + Send {
+    /// Handles an HTTP request.
+    fn handle_request(&self, req: &Request, arg: T) -> Response;
+}
+
+/// An HTTP routes structure.
+pub struct HttpRoutes<T> {
+    server_id: String,
+    prefix: String,
+    media_type: MediaType,
+    /// routes is a hash table mapping endpoint URIs to their endpoint handlers.
+    routes: HashMap<String, Box<dyn EndpointHandler<T> + Sync + Send>>,
+}
+
+impl<T: Send> HttpRoutes<T> {
+    /// Create a http request router.
+    pub fn new(server_id: String, prefix: String) -> Self {
+        HttpRoutes {
+            server_id,
+            prefix,
+            media_type: MediaType::ApplicationJson,
+            routes: HashMap::new(),
+        }
+    }
+
+    /// Register a request handler for a path.
+    #[allow(clippy::map_entry)]
```

Comment:
The pattern used in rust for lookup insert is with using `entry`. You can get rid of this allow and it is theoretically more efficient:
```rust
        let full = format!("{}:{}{}", method.to_str(), self.prefix, path);
        match self.routes.entry(full.clone()) {
            Occupied(_) => return Err(RouteError::HandlerExist(full)),
            Vacant(entry) => entry.insert(handler)
        };

        Ok(())
```

nit: can you also rename full to `full_path`?